src/values/AbstractValue.js

```diff
@@ -729,6 +730,7 @@ export default class AbstractValue extends Value {
     let labels = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
     invariant(labels.length >= operands.length);
     let operationDescriptor = createOperationDescriptor("ABSTRACT_FROM_TEMPLATE", { template });
+    if (resultType === ObjectValue || resultType === ArrayValue) hash = ++realm.objectCount;
```

Comment:
What about `FunctionValue`? Do you want `Value.isTypeCompatibleWith(resultType, ObjectValue)` that's already mentioned above?

In itself, an object result doesn't mean that the function isn't pure, this is just being conservative. Leave behind a comment that mentions this (future optimization opportunity).